cmd/wfnconvert/wfnconvert.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/csv"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	var o options
+	o.addFlags()
+	flag.Usage = func() {
+		fmt.Fprintln(os.Stderr, "wfnconvert provides various manipulations to WFN attributes")
+		fmt.Fprintln(os.Stderr, "usage: wfnconvert [options]")
+		fmt.Fprintln(os.Stderr, "options:")
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+	flag.Parse()
+	if err := o.validate(); err != nil {
+		fmt.Fprintf(os.Stderr, "wfnconvert: %v\n\n", err)
+		flag.Usage()
+	}
+	if err := wfnconvert(os.Stdin, os.Stdout, &o); err != nil {
+		fmt.Fprintf(os.Stderr, "wfnconvert: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func wfnconvert(in io.Reader, out io.Writer, o *options) error {
+	inCh := make(chan []string)
+	outCh := make(chan []string)
+
+	g, ctx := errgroup.WithContext(context.Background())
+	var inFunc, procFunc, outFunc func() error
+	if len(o.csvFields) == 0 {
+		inFunc = readLines(ctx, in, inCh)
+		procFunc = processLines(o, inCh, outCh)
+		outFunc = writeLines(out, outCh)
+	} else {
+		inFunc = readCSV(ctx, in, o.csvComma, inCh)
+		procFunc = processCSV(o, inCh, outCh)
+		outFunc = writeCSV(out, o.csvComma, outCh)
```

Comment:
This seems,.... weird
Why not create an interface with Read, Process, Write and two implementations: one for csv one for lines?
CSV implementation could probably just wrap the lines one, do splitting/joining and wrap the functions... just saying
Won't block on this, but I think it could be nicer that way

Reply:
@skogtwin yes, no?

Reply:
No. I'm at that stage of my life where I really like closures.

Reply:
why do only readers get context?

Reply:
Well, there's two possible race conditions here:
1) processor exits with error, inCh is still open, reader blocks on it forever
2) writer exits with error, outCh is still open, processor blocks forever
Everything else leads to proper termination through deferred `close()` on channels and rage-loops.
So, technically, processor should also get context to avoid (2), but writer just pumps data from channel to stdout and is very unlikely to fail, and range over channel reads so much nicer, than infinite loop with select in it... So I got lazy I guess.

Reply:
Added context to process() anyway. But you can see what I meant saying "reads so much nicer"

Reply:
yes, I agree writers don't need the context, but processors do.